cmd/login.go

```diff
@@ -102,208 +112,157 @@ type deviceTokenResponse struct {
 		}
 
 		// Access token has expired, we have to refresh it
+		authProvider := getLoginInformation(ctxt)
 		authProvider.grantType = "refresh_token"
 
 		if (authProvider.TokenURL != "") && (authProvider.ClientID != "") {
-
-			response, err := http.PostForm(authProvider.TokenURL,
-				url.Values{"grant_type": {authProvider.grantType},
-					"client_id":     {authProvider.ClientID},
-					"refresh_token": {ctxt.RefreshToken}})
-
-			if err != nil {
-				return "", fmt.Errorf("Cannot refresh access token - %s", err)
+			params := url.Values{
+				"refresh_token": {ctxt.RefreshToken},
 			}
-
-			var tokenResponse deviceTokenResponse
-			jsonDecoder := json.NewDecoder(response.Body)
-			if err := jsonDecoder.Decode(&tokenResponse); err != nil {
-				return "", fmt.Errorf("Cannot decode token response - %s", err)
+			tokenResponse := getTokenResponse(authProvider, params, ctxt, false)
+			if tokenResponse.ErrorString != "" {
+				logger.Warn("tokenResponse", log.String("error", tokenResponse.ErrorString))
+				cobra.CheckErr("oauth: Unexpected error from authentication provider")
 			}
 
-			switch tokenResponse.ErrorString {
-			case "authorization_pending":
-				// No op - we're waiting on the user to open the link and login
-			case "expired_token":
-				return "", fmt.Errorf("The login process was not completed in time - please login again")
-			case "access_denied":
-				return "", fmt.Errorf("Could not login - access was denied")
-			case "invalid_grant":
-				return "", fmt.Errorf("Could not login - expired credentials. Please use the login command to refresh your credentials")
-			case "":
-				// No Errors:
-				ctxt.AccessToken = tokenResponse.AccessToken
-				// Add a 10 second buffer to expiry to account for differences in clock time between client
-				// server and message transport time (oauth2 library does the same thing)
-				ctxt.AccessTokenExpiry = time.Now().Add(time.Second * time.Duration(tokenResponse.ExpiresIn-10))
-
-				// We also get an updated ID token, let's make sure we have the latest info
-				ParseIDToken(&tokenResponse, ctxt, authProvider.JwksURL)
-
-				fmt.Println(fmt.Sprintf("Successfully acquired new access token. Expiry: %s", ctxt.AccessTokenExpiry))
-
-				SetContext(ctxt, true)
-			}
+			ctxt.AccessToken = tokenResponse.AccessToken
+			ctxt.RefreshToken = tokenResponse.RefreshToken
+			// Add a 10 second buffer to expiry to account for differences in clock time between client
+			// server and message transport time (oauth2 library does the same thing)
+			ctxt.AccessTokenExpiry = time.Now().Add(time.Second * time.Duration(tokenResponse.ExpiresIn-10))
 
+			// We also get an updated ID token, let's make sure we have the latest info
+			ParseIDToken(&tokenResponse, ctxt, authProvider.JwksURL)
+			SetContext(ctxt, true)
+			fmt.Printf("Successfully acquired new access token. Expiry: %s", ctxt.AccessTokenExpiry.Format(time.RFC822))
```

Comment:
I left the comment, but removed the time. I think it is still useful to see a positive printout as we switched to interacting with a web server in the meantime and this will tell us that everything went well. I do agree with the date being confusing